File: Rulesage.Common/Repositories/Abstractions/IOperationRepository.fs
namespace Rulesage.Common.Repositories.Abstractions

open System.Runtime.InteropServices
open System.Threading
open System.Threading.Tasks
open Rulesage.Common.Types.Domain

type IOperationRepository =
    abstract member FindByIdAsync:
        id: int * [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<OperationBlueprint option>

    abstract member FindOrderByCosineDistance:
        queryVector: float32 array *
        take: int *
        [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<(OperationSignature * float32) array>

File: Rulesage.Shared/Repositories/Abstractions/IOperationRepository.cs
using Rulesage.Common.Types.Domain;

namespace Rulesage.Shared.Repositories.Abstractions;

public interface IOperationRepository : IDocumentRepository
{
    Task<OperationBlueprint?> FindByIdAsync(int id, CancellationToken cancellationToken = default);

    Task<IEnumerable<(OperationSignature, float)>> FindOrderByCosineDistanceAsync(float[] queryVector, int skip, int take,
        CancellationToken cancellationToken = default);
}